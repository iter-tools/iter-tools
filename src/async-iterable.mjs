async function * asyncify (iterable) {
  yield * iterable
}

const emptyArray = []

export default function asyncIterable (asyncIterator) {
  if (asyncIterator == null) {
    return asyncify(emptyArray)
  } else if (asyncIterator[Symbol.asyncIterator]) {
    return asyncIterator
  } else if (asyncIterator[Symbol.iterator]) {
    asyncIterator = asyncify(asyncIterator)[Symbol.asyncIterator]()
  } else if (typeof asyncIterator !== 'object' || typeof asyncIterator.next !== 'function') {
    throw new Error(`Expected to receive an async iterator of the form {next()}, but instead received: ${asyncIterator}`)
  }

  return {
    [Symbol.asyncIterator] () {
      return asyncIterator
    }
  }
}
