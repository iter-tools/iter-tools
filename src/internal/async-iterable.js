import { isAsyncIterable } from '../impls/is-async-iterable/is-async-iterable.js';
import { isAsyncLoopable } from '../impls/is-async-loopable/is-async-loopable.js';
import { isAsyncWrappable } from '../impls/is-async-wrappable/is-async-wrappable.js';
import { asyncNullableWrap as asyncWrap } from './async-wrap.js';
import { BaseResultIterable, ResultIterable } from './iterable.js';
import { variadicCurryWithValidation } from './curry.js';
import { _, __iterate } from './symbols.js';

export {
  asyncWrap,
  isAsyncIterable as asyncIsIterable,
  isAsyncLoopable as asyncIsLoopable,
  isAsyncWrappable as asyncIsWrappable,
};

export async function asyncCallReturn(iterator) {
  if ('return' in iterator) await iterator.return();
}

export function asyncEnsureIterable(value) {
  if (!isAsyncWrappable(value)) {
    if (typeof value.next === 'function') {
      throw new TypeError(
        'iter-tools received a value that looked like an iterator but was not async iterable. Get help fixing this: https://github.com/iter-tools/iter-tools/wiki/Making-iterators-iterable#async-iterators',
      );
    } else throw new TypeError('Expected an async iterable, sync iterable, null or undefined');
  } else if (isAsyncIterable(value)) {
    return value;
  } else {
    return asyncWrap(value);
  }
}

export function AsyncResultIterable(...args) {
  BaseResultIterable.apply(this, args);
}

AsyncResultIterable.prototype = Object.assign(Object.create(BaseResultIterable.prototype), {
  constructor: AsyncResultIterable,
  [Symbol.asyncIterator]() {
    return this[__iterate]();
  },
});

function AsyncSimpleResultIterable(...args) {
  AsyncResultIterable.apply(this, args);
}

AsyncSimpleResultIterable.prototype = Object.assign(Object.create(AsyncResultIterable.prototype), {
  [__iterate]() {
    return this[_].fn(...this[_].args);
  },
});

function makeFunctionConfig(fn, fnConfig = {}) {
  const {
    validateArgs,
    variadic,
    reduces,
    optionalArgsAtEnd,
    minArgs,
    maxArgs,
    forceSync,
  } = fnConfig;

  return {
    fn,
    validateArgs: validateArgs || ((_) => {}),
    variadic: !!variadic,
    reduces: !!reduces,
    optionalArgsAtEnd: !!optionalArgsAtEnd,
    minArgs: minArgs === undefined ? fn.length - 1 : minArgs,
    maxArgs: maxArgs === undefined ? fn.length - 1 : maxArgs,
    isIterable: isAsyncWrappable,
    iterableType: 'asyncIterable',
    applyOnIterableArgs: asyncEnsureIterable,
    IterableClass: forceSync ? ResultIterable : AsyncResultIterable,
  };
}

export async function asyncCache(it) {
  const arr = [];
  for await (const value of it) arr.push(value);
  return asyncWrapWithResultIterable(asyncWrap)(arr);
}

export function asyncWrapWithResultIterable(fn, { validateArgs = (_) => _ } = {}) {
  return (...args) => {
    validateArgs(args);
    return new AsyncSimpleResultIterable(fn, args);
  };
}

export const asyncIterableCurry = (fn, config) => {
  return variadicCurryWithValidation(makeFunctionConfig(fn, config));
};
