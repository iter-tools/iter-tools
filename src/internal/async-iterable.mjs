import { ensureIterable, isValidIterableArgument } from './iterable'
import { variadicCurryWithValidation } from './curry'

export function isAsyncIterable (i) {
  return Boolean(i != null && i[Symbol.asyncIterator])
}

export async function * asyncify (iterable) {
  if (isAsyncIterable(iterable)) {
    yield * iterable
  } else {
    // it should be enough "yield * iterable", but it is broken with es5 version
    for (const item of iterable) {
      yield Promise.resolve(item)
    }
  }
}

export function ensureAsyncIterable (i) {
  if (isAsyncIterable(i)) {
    return i
  } else {
    return asyncify(ensureIterable(i))
  }
}

export function isValidAsyncIterableArgument (i) {
  return isAsyncIterable(i) || isValidIterableArgument(i)
}

export const asyncIterableCurry = (fn, { variadic = false, reduces = false, forceSync = false } = {}, minArgs, maxArgs) => {
  return variadicCurryWithValidation(
    isValidAsyncIterableArgument,
    'asyncIterable',
    ensureAsyncIterable,
    fn,
    variadic,
    reduces,
    forceSync,
    minArgs,
    maxArgs
  )
}
