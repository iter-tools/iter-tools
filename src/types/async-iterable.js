// Placeholder to keep typed js tests from crashing
