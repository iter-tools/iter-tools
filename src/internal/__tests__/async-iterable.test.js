/**
 * @generated-from ./$iterable.test.js
 * This file is autogenerated from a template. Please do not edit it directly.
 * To rebuild it from its template use the command
 * > npm run generate
 * More information can be found in CONTRIBUTING.md
 */

import { asyncEnsureIterable, asyncIterableCurry } from '../async-iterable.js';
import { asyncWrap, asyncUnwrap } from '../../test/async-helpers.js';

describe('asyncEnsureIterable', () => {
  describe('when i is a sync iterable', () => {
    it('returns i wrapped in an async iterable', async () => {
      const i = asyncEnsureIterable([1, 2, 3]);
      expect(typeof i[Symbol.asyncIterator]).toBe('function');
      expect(await asyncUnwrap(asyncEnsureIterable(i))).toEqual([1, 2, 3]);
    });
  });

  describe(`when i is ${'asyncIterable'}`, () => {
    it('returns i', () => {
      const i = asyncWrap([1, 2, 3]);
      expect(i).toBe(asyncEnsureIterable(i));
    });
  });

  it('works with iterables', async () => {
    expect(await asyncUnwrap(asyncEnsureIterable(asyncWrap([1, 2, 3])))).toEqual([1, 2, 3]);
  });

  describe('when i is null', () => {
    it('returns the empty iterable', async () => {
      const i = asyncEnsureIterable(null);
      expect(await asyncUnwrap(i)).toEqual([]);
    });
  });

  describe('when i cannot be coerced to an iterable', () => {
    it('throws', () => {
      expect(() => asyncEnsureIterable(false)).toThrowErrorMatchingSnapshot();
    });
  });

  describe('when i looks like an iterator', () => {
    it('throws a more helpful error', () => {
      expect(() => asyncEnsureIterable({ next() {} })).toThrowErrorMatchingSnapshot();
    });
  });
});

class Hello {}
class Goodbye {}
class World {}

const hello = new Hello();
const goodbye = new Goodbye();
const world = new World();

async function* iter(...args) {
  yield* args;
}

async function add(initial, iterable) {
  let result = initial;
  for await (const number of iterable) {
    result += number;
  }
  return result;
}

async function addAll(initial, iterables) {
  let result = initial;
  for (const iterable of iterables) {
    for await (const number of iterable) {
      result += number;
    }
  }
  return result;
}

describe('asyncIterableCurry', () => {
  const f2 = (iterable, a, b) => iter(a, b);
  const f1 = (iterable, a) => iter(a);
  const f0 = (iterable) => iter();
  const c2 = asyncIterableCurry(f2);
  const c1 = asyncIterableCurry(f1);
  const c0 = asyncIterableCurry(f0);
  /* eslint-disable no-unused-expressions */
  f2.name; // Make sure names don't get thrown away by babel-minify
  f1.name;
  f0.name;
  /* eslint-enable no-unused-expressions */

  it('curries', async () => {
    expect(await asyncUnwrap(c2(hello, world, []))).toEqual([hello, world]);
    expect(await asyncUnwrap(c2(hello, world)([]))).toEqual([hello, world]);
    expect(await asyncUnwrap(c2(hello)(world, []))).toEqual([hello, world]);
    expect(await asyncUnwrap(c2(hello)(world)([]))).toEqual([hello, world]);
    expect(await asyncUnwrap(c1(hello, []))).toEqual([hello]);
    expect(await asyncUnwrap(c1(hello)([]))).toEqual([hello]);
    expect(await asyncUnwrap(c0([]))).toEqual([]);
  });

  it('ignores extra arguments after iterable', async () => {
    expect(await asyncUnwrap(c2(hello, world, [], 'foo'))).toEqual([hello, world]);
    expect(await asyncUnwrap(c1(hello)([], null))).toEqual([hello]);
    expect(await asyncUnwrap(c0([], 4))).toEqual([]);
  });

  it('creates a static iterator', async () => {
    const c = asyncIterableCurry(async function* $wrap(i) {
      yield* i;
    });
    expect(await c([1]).next()).toEqual({ value: 1, done: false });
    expect(await c([]).return()).toEqual({ value: undefined, done: true });
    expect(
      await (async () => {
        try {
          await c([]).throw(new Error());
        } catch (e) {
          return e;
        }
      })(),
    ).toBeInstanceOf(Error);
  });

  it('throws with empty invocations', () => {
    expect(() => c2()(hello)(world)([])).toThrowErrorMatchingSnapshot();
    expect(() => c2(hello)()(world)([])).toThrowErrorMatchingSnapshot();
    expect(() => c2(hello)(world)()([])).toThrowErrorMatchingSnapshot();
  });

  it('throws with too many args', () => {
    expect(() => c2(hello)(goodbye)(world)([])).toThrowErrorMatchingSnapshot();
    expect(() => c1(hello)(world)([])).toThrowErrorMatchingSnapshot();
    expect(() => c0(hello)([])).toThrowErrorMatchingSnapshot();
  });

  describe('validates args', () => {
    it('can stop method execution by throwing an error', () => {
      const helloImpl = jest.fn();
      const hello = asyncIterableCurry(helloImpl, {
        minArgs: 1,
        maxArgs: 1,
        validateArgs(args) {
          const [world] = args;
          if (!(world instanceof World)) {
            throw new Error('Expected the world');
          }
        },
      });
      expect(() => hello(null, [])).toThrowErrorMatchingSnapshot();
      expect(helloImpl).not.toHaveBeenCalled();
    });

    it('can alter arguments', async () => {
      const empty = (async function* () {})();
      const helloImpl = jest.fn(async function* () {});
      const hello = asyncIterableCurry(helloImpl, {
        minArgs: 1,
        maxArgs: 1,
        validateArgs(args) {
          args[0] = world;
        },
      });
      await asyncUnwrap(hello(null, empty));
      expect(helloImpl).toHaveBeenCalledWith(empty, world);
    });
  });

  describe('when passed explicit arity', () => {
    const f = (c, a = goodbye, b = world) => iter(a, b);
    const c = asyncIterableCurry(f, { minArgs: 0, maxArgs: 2 });
    /* eslint-disable no-unused-expressions */
    f.name; // Make sure it don't get thrown away by babel-minify
    /* eslint-enable no-unused-expressions */

    it('curries', async () => {
      expect(await asyncUnwrap(c(hello)(world)([]))).toEqual([hello, world]);
      expect(await asyncUnwrap(c(hello)([]))).toEqual([hello, world]);
      expect(await asyncUnwrap(c([]))).toEqual([goodbye, world]);
    });

    it('throws with empty invocations', () => {
      expect(() => c2()(hello)(world)([])).toThrowErrorMatchingSnapshot();
    });

    it('throws with too many args', () => {
      expect(() => c(hello)(goodbye)(world)([])).toThrow(
        new Error(
          `f takes up to 2 arguments, followed by ${'asyncIterable'}. You already passed 3 arguments and the last argument was not ${'asyncIterable'}`,
        ),
      );
    });
  });

  describe('works with reducing functions', () => {
    const f2 = (iterable, a, b) => add(a + b, iterable);
    const f1 = (iterable, a) => add(a, iterable);
    const f0 = (iterable) => add(0, iterable);
    const c2 = asyncIterableCurry(f2, { reduces: true });
    const c1 = asyncIterableCurry(f1, { reduces: true });
    const c0 = asyncIterableCurry(f0, { reduces: true });
    /* eslint-disable no-unused-expressions */
    f2.name; // Make sure names don't get thrown away by babel-minify
    f1.name;
    f0.name;
    /* eslint-enable no-unused-expressions */

    it('curries', async () => {
      expect(await c2(1)(2)([4])).toBe(7);
      expect(await c1(1)([2])).toBe(3);
      expect(await c0([1])).toBe(1);
    });

    it('throws with empty invocations', () => {
      expect(() => c2()(1, 2, [4])).toThrowErrorMatchingSnapshot();
      expect(() => c2(1, 2)()([4])).toThrowErrorMatchingSnapshot();
    });

    it('throws with too many args', () => {
      expect(() => c2(1)(2)(3)([])).toThrowErrorMatchingSnapshot();
      expect(() => c1(1)(2)([])).toThrowErrorMatchingSnapshot();
      expect(() => c0(1)([])).toThrowErrorMatchingSnapshot();
    });
  });

  describe('works with variadic functions', () => {
    const f1 = (iterables, a) => addAll(a, iterables);
    /* eslint-disable no-unused-expressions */
    f1.name; // Make sure it doesn't get thrown away by babel-minify
    /* eslint-enable no-unused-expressions */
    const c1 = asyncIterableCurry(f1, { variadic: true, reduces: true });

    it('curries', async () => {
      expect(await c1(1)([2, 4], [8, 16])).toBe(31);
      expect(await c1(1)([2, 4])).toBe(7);
    });

    it('throws with empty invocations', () => {
      expect(() => c1(1)()([2, 4])).toThrowErrorMatchingSnapshot();
    });

    it('throws with too many args', () => {
      expect(() => c1(1)(2)([4, 8])).toThrowErrorMatchingSnapshot();
    });
  });
});
