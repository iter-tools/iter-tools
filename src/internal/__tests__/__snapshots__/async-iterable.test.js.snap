// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asyncEnsureIterable when i cannot be coerced to an iterable throws 1`] = `"Expected an async iterable, sync iterable, null or undefined"`;

exports[`asyncEnsureIterable when i looks like an iterator throws a more helpful error 1`] = `"iter-tools received a value that looked like an iterator but was not async iterable. Get help fixing this: https://github.com/iter-tools/iter-tools/wiki/Making-iterators-iterable#async-iterators"`;

exports[`asyncIterableCurry throws with empty invocations 1`] = `"Cannot make a partial application with no arguments"`;

exports[`asyncIterableCurry throws with empty invocations 2`] = `"Cannot make a partial application with no arguments"`;

exports[`asyncIterableCurry throws with empty invocations 3`] = `"Cannot make a partial application with no arguments"`;

exports[`asyncIterableCurry throws with too many args 1`] = `"f2 takes up to 2 arguments, followed by asyncIterable. You already passed 3 arguments and the last argument was not asyncIterable"`;

exports[`asyncIterableCurry throws with too many args 2`] = `"f1 takes up to 1 arguments, followed by asyncIterable. You already passed 2 arguments and the last argument was not asyncIterable"`;

exports[`asyncIterableCurry throws with too many args 3`] = `"f0 takes up to 0 arguments, followed by asyncIterable. You already passed 1 arguments and the last argument was not asyncIterable"`;

exports[`asyncIterableCurry validates args can stop method execution by throwing an error 1`] = `"Expected the world"`;

exports[`asyncIterableCurry when passed explicit arity throws with empty invocations 1`] = `"Cannot make a partial application with no arguments"`;

exports[`asyncIterableCurry works with reducing functions throws with empty invocations 1`] = `"Cannot make a partial application with no arguments"`;

exports[`asyncIterableCurry works with reducing functions throws with empty invocations 2`] = `"Cannot make a partial application with no arguments"`;

exports[`asyncIterableCurry works with reducing functions throws with too many args 1`] = `"f2 takes up to 2 arguments, followed by asyncIterable. You already passed 3 arguments and the last argument was not asyncIterable"`;

exports[`asyncIterableCurry works with reducing functions throws with too many args 2`] = `"f1 takes up to 1 arguments, followed by asyncIterable. You already passed 2 arguments and the last argument was not asyncIterable"`;

exports[`asyncIterableCurry works with reducing functions throws with too many args 3`] = `"f0 takes up to 0 arguments, followed by asyncIterable. You already passed 1 arguments and the last argument was not asyncIterable"`;

exports[`asyncIterableCurry works with variadic functions throws with empty invocations 1`] = `"Cannot make a partial application with no arguments"`;

exports[`asyncIterableCurry works with variadic functions throws with too many args 1`] = `"f1 takes up to 1 arguments, followed by ...asyncIterables. You already passed 2 arguments and the following arguments were not all asyncIterables"`;
